test/e2e/storage/storageclass.go

```diff
@@ -0,0 +1,161 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	types "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+	admissionapi "k8s.io/pod-security-admission/api"
+
+	"github.com/onsi/ginkgo/v2"
+	"github.com/onsi/gomega"
+)
+
+var _ = utils.SIGDescribe("StorageClasses", func() {
+
+	f := framework.NewDefaultFramework("csi-storageclass")
+	f.NamespacePodSecurityLevel = admissionapi.LevelBaseline
+
+	ginkgo.Describe("CSI Conformance", func() {
+		ginkgo.It("should run through the lifecycle of a StorageClass", func(ctx context.Context) {
+
+			scClient := f.ClientSet.StorageV1().StorageClasses()
+			var initialSC, replacementSC *storagev1.StorageClass
+
+			initialSC = &storagev1.StorageClass{
+				TypeMeta: metav1.TypeMeta{
+					Kind: "StorageClass",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "e2e-",
+				},
+				Provisioner: "e2e-fake-provisioner",
+			}
+
+			ginkgo.By("Creating a StorageClass")
+			createdStorageClass, err := scClient.Create(ctx, initialSC, metav1.CreateOptions{})
+			framework.ExpectNoError(err, "failed to create the requested StorageClass")
+
+			ginkgo.By(fmt.Sprintf("Get StorageClass %q", createdStorageClass.Name))
+			retrievedStorageClass, err := scClient.Get(ctx, createdStorageClass.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get StorageClass %q", createdStorageClass.Name)
+
+			ginkgo.By(fmt.Sprintf("Patching the StorageClass %q", retrievedStorageClass.Name))
+			payload := "{\"metadata\":{\"labels\":{\"" + retrievedStorageClass.Name + "\":\"patched\"}}}"
+			patchedStorageClass, err := scClient.Patch(ctx, retrievedStorageClass.Name, types.StrategicMergePatchType, []byte(payload), metav1.PatchOptions{})
+			framework.ExpectNoError(err, "failed to patch StorageClass %q", retrievedStorageClass.Name)
+			gomega.Expect(patchedStorageClass.Labels).To(gomega.HaveKeyWithValue(patchedStorageClass.Name, "patched"), "checking that patched label has been applied")
+
+			ginkgo.By(fmt.Sprintf("Delete StorageClass %q", patchedStorageClass.Name))
+			err = scClient.Delete(ctx, patchedStorageClass.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "failed to delete StorageClass %q", patchedStorageClass.Name)
+
+			ginkgo.By(fmt.Sprintf("Confirm deletion of StorageClass %q", patchedStorageClass.Name))
+
+			scSelector := labels.Set{patchedStorageClass.Name: "patched"}.AsSelector().String()
+			type state struct {
+				StorageClasses []storagev1.StorageClass
+			}
+
+			err = framework.Gomega().Eventually(ctx, framework.HandleRetry(func(ctx context.Context) (*state, error) {
```

Comment:
Hi @pohly, I have a question what's difference between`framework.Gomega().Eventually` and `gomega.Eventually`. How to choose them when writing test cases? 

Can you help me to clarify it? Thanks in advance.

Reply:
This is explained in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/writing-good-e2e-tests.md#polling-and-timeouts under "The E2E framework therefore uses a different approach".

The gist of it is that top-level tests can use `gomega.Eventually`, but helper functions should use `framework.Gomega().Eventually` and return an error, potentially with a wrapped error ("%w", not "%v") to add more context to the failure.